src/Adapter/Product/Stock/Repository/StockAvailableMultiShopRepository.php

```diff
@@ -233,6 +248,76 @@ public function getAssociatedShopIds(StockId $stockId): array
         return $shops;
     }
 
+    /**
+     * Updates the physical_quantity and reserved_quantity columns for the specified Stock. Most of this function logic comes from
+     * StockManager::updatePhysicalProductQuantity
+     *
+     * @param StockId $stockId
+     * @param OrderStateId $errorStateId
+     * @param OrderStateId $canceledStateId
+     */
+    public function updatePhysicalProductQuantity(StockId $stockId, OrderStateId $errorStateId, OrderStateId $canceledStateId): void
+    {
+        $this->updateReservedProductQuantity($stockId, $errorStateId, $canceledStateId);
+
+        // Now update the physical_quantity
+        $updateQb = $this->connection->createQueryBuilder();
+        $updateQb
+            ->update($this->dbPrefix . 'stock_available', 'sa')
+            ->set('physical_quantity', 'sa.quantity + sa.reserved_quantity')
+            ->where('sa.id_stock_available = :stockId')
+            ->setParameter('stockId', $stockId->getValue())
+        ;
+        $updateQb->execute();
+    }
+
+    protected function updateReservedProductQuantity(StockId $stockId, OrderStateId $errorStateId, OrderStateId $canceledStateId): void
+    {
+        $qb = $this->connection->createQueryBuilder();
+        $qb
+            ->addSelect('SUM(od.product_quantity - od.product_quantity_refunded) AS reserved_quantity')
+            ->from($this->dbPrefix . 'orders', 'o')
+            ->innerJoin('o', $this->dbPrefix . 'order_detail', 'od', 'od.id_order = o.id_order')
+            ->innerJoin('o', $this->dbPrefix . 'order_state', 'os', 'os.id_order_state = o.current_state')
+            ->innerJoin(
+                'od', $this->dbPrefix . 'stock_available', 'sa',
+                'od.product_id = sa.id_product AND od.product_attribute_id = sa.id_product_attribute AND od.id_shop = sa.id_shop'
+            )
+            ->where($qb->expr()->and(
+                $qb->expr()->eq('o.id_shop', 'sa.id_shop'),
+                $qb->expr()->neq('os.shipped', 1),
+                $qb->expr()->or(
+                    $qb->expr()->eq('o.valid', 1),
+                    $qb->expr()->and(
+                        $qb->expr()->neq('os.id_order_state', ':errorStateId'),
+                        $qb->expr()->neq('os.id_order_state', ':canceledStateId')
+                    )
+                ),
+                $qb->expr()->eq('sa.id_stock_available', ':stockId')
+            ))
+            ->groupBy('od.product_id', 'od.product_attribute_id')
+            ->setParameters([
+                'stockId' => $stockId->getValue(),
+                'errorStateId' => $errorStateId->getValue(),
+                'canceledStateId' => $canceledStateId->getValue(),
+            ])
+        ;
+
```

Comment:
Put this code in a `AbstractStockAvailableRepository` to avoid code duplicate with `updateReservedProductQuantity::StockAvailableRepository`

Reply:
I wish I could but it's not that easy because `StockAvailableRepository` and `StockAvailableMultiShopRepository` don't have the same parent (`AbstractObjectModelRepository` and `AbstractMultiShopObjectModelRepository` for the other) so I can't create a common abstract class.

The reason why we have those two different repositories is that the multishop development is not finished everywhere on the product page, the strategy was to handle it feature by feature. For stock management it is done for regular product but not for combination for example that's why we introduce some temporary multishop repositories which are indeed doublons with the standard initial ones.

We need both right now because they don't have the same interfaces (multishop repositories force the use of a ShopConstraint parameter), the idea is to one day only have one of the two when everything is done and merge them together but sadly we didn't have time to finish the multishop dev everywhere so we have to keep those two different repositories for now.

Reply:
Since the multishop will likely be finished in v8.1 it also means that we will have to keep both services to avoid a breaking change, so the two repositories will be merged into one only in v9 now I guess.

That's why for now we have indeed duplicate code between the multishop repositories and the regular ones (the same problem applies in other domains with multishop repositories). Except by using traits I don't see how we can fix this, but many maintainers are not comfortable with use of traits and they would also be trait bound to be removed at some point which will also imply a breaking change. So no perfect solution here.

Reply:
Just for the record I ❤️ traits 😛 

But it seems it's not everybody 😄